extensions/bundle/main.ts

```diff
@@ -0,0 +1,266 @@
+import {
+  green,
+  red,
+  yellow,
+} from "https://deno.land/std@0.150.0/fmt/colors.ts";
+import { PhylumApi, PolicyEvaluationResponseRaw } from "phylum";
+
+class FileBackup {
+  readonly fileName: string;
+  fileContent: string | null;
+
+  constructor(fileName: string) {
+    this.fileName = fileName;
+    this.fileContent = null;
+  }
+
+  async backup() {
+    try {
+      this.fileContent = await Deno.readTextFile(this.fileName);
+    } catch (_e) { /* Do nothing */ }
+  }
+
+  async restoreOrDelete() {
+    try {
+      if (this.fileContent != null) {
+        await Deno.writeTextFile(this.fileName, this.fileContent);
+      } else {
+        await Deno.remove(this.fileName);
+      }
+    } catch (_e) { /* Do nothing */ }
+  }
+}
+
+// Find project root directory.
+async function findRoot(manifest: string): Promise<string | undefined> {
+  let workingDir = Deno.cwd();
+
+  // Traverse up to 32 directories to find the root directory.
+  for (let i = 0; i < 32; i++) {
+    try {
+      // Check if manifest exists at location.
+      await Deno.stat(workingDir + "/" + manifest);
+      return workingDir;
+    } catch (_e) {
+      // Pop to parent if manifest doesn't exist.
+      workingDir += "/..";
+    }
+  }
+
+  return undefined;
+}
+
+// Ensure no arguments are passed before a subcommand.
+const firstSubcommand = Deno.args.findIndex((arg) => !arg.startsWith("-"));
+if (firstSubcommand > 0) {
+  console.error(
+    `[${
+      red("phylum")
+    }] This extension does not support arguments before the first subcommand. Please open an issue if "${
+      Deno.args[0]
+    }" is not an argument.`,
+  );
+  Deno.exit(127);
+}
+
+// Ignore all commands that shouldn't be intercepted.
+if (
+  Deno.args.length == 0 ||
+  !(
+    "install".startsWith(Deno.args[0]) ||
+    "update".startsWith(Deno.args[0]) ||
+    "add".startsWith(Deno.args[0])
+  )
```

Comment:
At that point the vulnerable package has already made it into the project though, so it's kinda too late anyway.

We don't do this for any of the ecosystem extensions, which I do think makes some sense. But it could be added in theory if we saw some value in it.

Reply:
> At that point the vulnerable package has already made it into the project though, so it's kinda too late anyway.

Not necessarily. The gem removed with the `remove` command is not the one that is the issue here. It is any other gem(s) (or their dependencies?) that exist in the `Gemfile`. They will get installed, if not already. This can happen when checking out a project and using the `remove` command prior to an `install`. It can also happen when modifying the `Gemfile` manually. Here is an example, with the known "bad" (due to license issue) gem `locale@2.1.3`:

```
# Prove that the `locale` gem is not installed
../rubocop  13  22 on  master via 💎 v3.2.2
❯ ~/dev/phylum/localdev/cli/target/debug/phylum bundle info locale
Could not find gem 'locale'.

../rubocop  13  22 on  master via 💎 v3.2.2
❌ 7 ❯ ls -alh /opt/homebrew/lib/ruby/gems/3.2.0/gems/locale*
zsh: no matches found: /opt/homebrew/lib/ruby/gems/3.2.0/gems/locale*

# Add the `locale` gem to the `Gemfile`
../rubocop  13  22 on  master via 💎 v3.2.2
❌ 1 ❯ vim Gemfile

../rubocop  13  22 on  master [!] via 💎 v3.2.2 took 53s
❯ git diff Gemfile
diff --git a/Gemfile b/Gemfile
index 20f333152..43ae11b93 100644
--- a/Gemfile
+++ b/Gemfile
@@ -4,6 +4,7 @@ source 'https://rubygems.org'

 gemspec

+gem 'locale', '2.1.3'
 gem 'asciidoctor'
 gem 'bump', require: false
 gem 'bundler', '>= 1.15.0', '< 3.0'

# Show that updating the `Gemfile` does not add/install the gem
../rubocop  13  22 on  master [!] via 💎 v3.2.2
❯ ~/dev/phylum/localdev/cli/target/debug/phylum bundle info locale
Could not find gem 'locale (= 2.1.3)' in locally installed gems.

# Show that using the extension to `remove` a gem will also `install` the Gemfile, to
# include any other gems that have been added to the Gemfile, outside of Phylum analysis.
../rubocop  13  22 on  master [!] via 💎 v3.2.2
❌ 7 ❯ ~/dev/phylum/localdev/cli/target/debug/phylum bundle remove asciidoctor
Removing gems from /Users/maxrake/dev/rubocop/Gemfile
asciidoctor was removed.
Fetching gem metadata from https://rubygems.org/..........
Resolving dependencies...
Fetching locale 2.1.3
Installing locale 2.1.3

../rubocop  13  22 on  master [!] via 💎 v3.2.2
❯ ~/dev/phylum/localdev/cli/target/debug/phylum bundle info locale
  * locale (2.1.3)
	Summary: Ruby-Locale is the pure ruby library which provides basic APIs for localization.
	Homepage: https://github.com/ruby-gettext/locale
	Path: /opt/homebrew/lib/ruby/gems/3.2.0/gems/locale-2.1.3
```

> We don't do this for any of the ecosystem extensions, which I do think makes some sense. But it could be added in theory if we saw some value in it.

Not all the ecosystems work the same way. Poetry, for example, has a `--lock` option to go with the `remove` subcommand so that the changes are only applied to the lockfile and no packages are added/removed/updated to the environment.

Making changes to the ecosystem extensions does seem worth doing to prevent this gap in coverage, but that could be a separate issue.

Reply:
https://github.com/phylum-dev/cli/issues/1189 was created for other default extensions and this PR was updated to include the `remove` command for coverage.